### PR TITLE
Migrate to the HsYAML package

### PR DESCRIPTION
**Pull request by vijayphoenix**
This PR address the issue #744 
Some other issues like #1148 can also be solved after this PR gets merged.

HsYAML-0.2 and HsYAML-aeson 0.2 will be released very soon.

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -1 +1,10 @@
 packages: ./dhall ./dhall-bash ./dhall-json ./dhall-lsp-server ./dhall-nix
+
+source-repository-package
```
Comment (vijayphoenix) [inline cabal.project]: HsYAML-0.2 and HsYAML-aeson 0.2 will be released very soon.

**Comment by jneira** [comment]
To make appveyor and travis builds work we'll should add the new dependencies to stack.yaml too

File: dhall-json/src/Dhall/Yaml.hs [hunk 2]
```diff
@@ -90,32 +94,54 @@ jsonToYaml
...
-    encodeYaml = Data.Yaml.encode
-# else
-    encodeYaml = Data.Yaml.encodeWith
-
-    customStyle = \s -> case () of
+    defaultSchemaEncoder = Y.setScalarStyle style Y.defaultSchemaEncoder
+
+    defaultEncodeStr s = case () of
+      ()
+        | "\n" `Text.isInfixOf` s -> Right (YE.untagged, YE.Literal YE.Clip YE.IndentAuto, s)
```
Comment (Gabriel439) [inline dhall-json/src/Dhall/Yaml.hs]: Thanks for taking the effort to preserve the old behavior!

File: dhall-json/src/Dhall/Yaml.hs [hunk 3]
```diff
@@ -90,32 +94,54 @@ jsonToYaml
...
+        else defaultSchemaEncoder
 #endif
+
+
+{-# INLINE bsToStrict #-}
+bsToStrict :: Data.ByteString.Lazy.ByteString -> ByteString
+#if MIN_VERSION_bytestring(0,10,0)
+bsToStrict = Data.ByteString.Lazy.toStrict
+#else
+bsToStrict = Data.ByteString. Data.ByteString.Lazy.toChunks
```
Comment (Gabriel439) [inline dhall-json/src/Dhall/Yaml.hs]: Is this what you meant?

```suggestion
bsToStrict = Data.ByteString.concat Data.ByteString.Lazy.toChunks
```

Comment (vijayphoenix) [reply dhall-json/src/Dhall/Yaml.hs]: yeah

**Comment by Gabriel439** [comment]
@vijayphoenix: Based on the Nix build failure:

https://hydra.dhall-lang.org/build/29423/nixlog/4

... it looks like `HsYAML` fails to build against `mtl-2.2.2` which introduced the `liftEither` utility.  You can either fix `HsYAML` to build against `mtl` and then update the build to reference that new revision or you can downgrade `mtl` to version 2.2.1 for the Nix build by running:

```
$ cabal2nix cabal://mtl-2.2.1 > ./nix/mtl.nix
```

File: cabal.project [hunk 4]
```diff
@@ -1 +1,10 @@
...
+    type: git
+    location: https://github.com/vijayphoenix/HsYAML.git
+    tag: 9354f53c69e9178a8e7c5c17b47137e23d4179df
+
+source-repository-package
+    type: git
+    location: https://github.com/vijayphoenix/HsYAML-aeson.git
```
Comment (quasicomputational) [inline cabal.project]: Can this get a commit id too, please?

File: dhall-json/dhall-json.cabal [hunk 5]
```diff
@@ -76,14 +72,10 @@ Library
...
-        Build-Depends:
-            yaml >= 0.5.0 && < 0.11
-      else
         Build-Depends:
-            libyaml >= 0.1.1.0 && < 0.2 ,
-            yaml    >= 0.11.0  && < 0.12
-
+            HsYAML >= 0.2,
+            HsYAML-aeson >= 0.2
+
```
Comment (quasicomputational) [inline dhall-json/dhall-json.cabal]: Trailing whitespace.

File: cabal.project (hunk 4 above)
Comment (vijayphoenix) [reply cabal.project]: b4808ee9bd7e73eeafc7f6695e98a8c591d68f11

**Comment by hvr** [comment]
@Gabriel439 the problem with mtl is a lot more tricky; it was basically a botched release; see https://github.com/haskell-hvr/HsYAML/issues/6 for details

**Comment by hvr** [comment]
@Gabriel439 I tried to reproduce this (but I don't have Nix, so the best I could do is to try with my own `ghcjs-8.4.0` packaging) but I didn't; I don't understand yet what Nix is doing different that breaks the code below

Basically the problem is the following fragment

```hs
{-# LANGUAGE CPP #-}

module Mtl222 (liftEither) where

#if !MIN_VERSION_mtl(2,2,2) || (__GLASGOW_HASKELL__ == 804 && __GLASGOW_HASKELL_PATCHLEVEL1__ < 2)
import           Control.Monad.Except         (MonadError (throwError))
#else
import           Control.Monad.Except         (liftEither)
#endif

-- GHC 8.4.1 shipped with a phony `mtl-2.2.2` and so we have to assume
-- pessimistically that on GHC 8.4.1 only, mtl-2.2.2 may be broken (even if it was reinstalled)
#if !MIN_VERSION_mtl(2,2,2) || (__GLASGOW_HASKELL__ == 804 && __GLASGOW_HASKELL_PATCHLEVEL1__ < 2)
#warning defining local liftEither
liftEither :: MonadError e m => Either e a -> m a
liftEither = either throwError return
#endif
```

I'm open to suggestions how to workaround the Nix issue; I'm starting to think that nobody shall ever use `liftEither`  as there's broken GHC versions out there that poison everything so that you can't trust the `MIN_VERSION_mtl(2,2,2)` macro *ever*

**Comment by jneira** [comment]
Appveyor builds it succesfully but fails at test step due to order of fields in yaml output:
```
./tasty/data/normal:           FAIL
    tasty\Main.hs:146:
    Conversion to YAML did not generate the expected output
    expected: "bool_value: true\ntext: |\n  Plain text\nstring_value: 2000-01-01\nint_value: 1\n"
     but got: "bool_value: true\nint_value: 1\nstring_value: 2000-01-01\ntext: |\n  Plain text\n"
  ./tasty/data/quoted:           FAIL (0.01s)
    tasty\Main.hs:146:
    Conversion to YAML did not generate the expected output
    expected: "'bool_value': true\n'text': |\n  Plain text\n'string_value': '2000-01-01'\n'int_value': 1\n"
     but got: "'bool_value': true\n'int_value': 1\n'string_value': '2000-01-01'\n'text': |\n  Plain text\n"
```
Not sure if we need to change the code or the tests (see #1187 for some context)

**Comment by Gabriel439** [comment]
@jneira @vijayphoenix: It's fine to update the failing test to match the new result

The issue here is that the old expected test result actually had the wrong order of fields, but the test still passed because the `yaml` library discards field ordering when decoding to JSON.  This is because `aeson`'s `Value` type doesn't provide a way to preserve field order.

Now that we're using `HsYAML` we're preserving the field order so the test suite is actually more accurate, and the fix is to update the expected output to correctly sort the fields.

**Comment by Gabriel439** [comment]
@vijayphoenix: Sorry, I gave you the wrong command for configuring `HsYAML-aeson` for Nix.  It should have been:

```
$ cabal2nix --revision b4808ee9bd7e73eeafc7f6695e98a8c591d68f11 https://github.com/haskell-hvr/HsYAML-aeson.git > ./nix/HsYAML-aeson.nix
```

Making that change will fix the `HsYAML-aeson` build failure for the Nix build

**Comment by vijayphoenix** [comment]
> @vijayphoenix: Sorry, I gave you the wrong command for configuring `HsYAML-aeson` for Nix. It should have been:
> 
> ```
> $ cabal2nix --revision b4808ee9bd7e73eeafc7f6695e98a8c591d68f11 https://github.com/haskell-hvr/HsYAML-aeson.git > ./nix/HsYAML-aeson.nix
> ```
> 
> Making that change will fix the `HsYAML-aeson` build failure for the Nix build

I think it should be 😉
```
$ cabal2nix --revision b4808ee9bd7e73eeafc7f6695e98a8c591d68f11 https://github.com/vijayphoenix/HsYAML-aeson.git > ./nix/HsYAML-aeson.nix
```

**Comment by Gabriel439** [comment]
@vijayphoenix: It looks like `HsYAML-aeson` depends on a newer version of `text`.  You can update to the latest version of `text` by running:

```
$ cabal2nix cabal://text > ./nix/text.nix
```

**Comment by hvr** [comment]
@Gabriel439,    @vijayphoenix workarounded this by allowing HsYAML-aeson to use text-1.2.3.0 (which is the version bundled with GHCJS 8.4.0 for which it probably isn't a good idea to force upgrading or reinstalling `text` anyway)

But now Hydra is hitting a different problem: https://hydra.dhall-lang.org/log/r7rzs5p561vz23crzqqcf1cplfzgk0ir-HsYAML-aeson-0.2.0.0.drv

My educated guess would be that it's trying to parse `HsYAML-aeson.cabal` (whose spec-version is 2.2) directly with a pre-2.0 lib:Cabal version which obviously cannot understand the newer 2.x spec-version formats (NB: newer lib:Cabal versions will refuse to even try to parse too new spec-versions)

How can this Hydra CI failure be fixed?

**Comment by Gabriel439** [comment]
@hvr: The reason it's hitting that error is because we also build `dhall-json` for GHC version 7.10 in CI, and for Nix that GHC version is tied to an older version of `Cabal`.  Upgrading `Cabal` for the older GHC version might be possible albeit tricky

It would probably be simpler to either (A) drop `dhall-json` support for GHC 7.10 (cc: @jneira who added GHC 7.10 support) or (B) see if `HsYAML-aeson` could support older versions of `Cabal`

**Comment by hvr** [comment]
@Gabriel439 Tbh, I'm a bit reluctant to downgrade cabal spec versions; we go to great length to support quite old legacy GHC versions in latest Cabal versions (Cabal 3.0 supports all GHCs back till GHC 7.0) so we can use current syntax in .cabal file; and note that Cabal 2.2 is already 2 major versions behind the latest spec-version 3.0, so it's not even considered "bleeding" edge. So I'd rather prefer there be a way for Nix to not tie the cabal spec version to GHC versions as otherwise this would render all our efforts with the goal to have users use the latest cabal spec-version regardless of targetted GHC version somewhat futile (at least from the POV of Nix-based Haskell development).

**Comment by sjakobi** [comment]
If we want to use a more recent GHC version on Hydra, we could probably move CI for dhall-json with GHC-7.10 to AppVeyor.

I do wonder how much longer we have to provide support for GHC-7.10 though…

**Comment by jneira** [comment]
In fact we dropped support of `dhall-json` for `ghc-7.10.3` some time ago: it was removed from `stack-lts-6.yaml` and `appveyor.yml`, so i think it could be removed from `hydra` too.

**Comment by Gabriel439** [comment]
Oh, in that case @vijayphoenix can comment out this line to drop Nix support for `dhall-json` on GHC 7.10:

https://github.com/dhall-lang/dhall-haskell/blob/06dc9ab55fbcc8d4a57f1719d0d0766c155e8a80/release.nix#L31

**Comment by sjakobi** [comment]
> In fact we dropped support of `dhall-json` for `ghc-7.10.3` some time ago: it was removed from `stack-lts-6.yaml` and `appveyor.yml`, so i think it could be removed from `hydra` too.

@jneira, if we don't need to support GHC-7.10 to keep `dhall-json` compatible with Eta, can we also remove support for GHC-7.10 in `dhall`?

**Comment by jneira** [comment]
@sjakobi I am afraid that eta and etlas are being built with ghc-7.10. As we added builtin support for dhall in etlas to make it its default configuration lang via a `dhall-to-cabal` fork we need to build dhall with ghc-7.10 as well (but no `dhall-json`, cause we don't use it).
Hopefully we will build `eta`/`etlas` with a newer ghc version at some point but there is no a concrete time limit.

**Comment by hvr** [comment]
Btw, Eta uses Etlas, which is a fork of Cabal which afaik supports newer spec-versions that the lib:Cabal 1.24 bundled with GHC 7.10 vanilla does (there's a growing number of packages on Hackage that started using 2.x spec versions that wouldn't  be accessible; so there'd be pressure for Eta to not be stuck on old lib:Cabals even thought they're a fork of the GHC 7.10 code-base).

I just tried the Eta docker image (`docker run --rm -it typelead/eta`) and it shows

```
# etlas --version
etlas version 1.6.0.0
compiled using version 2.1.0.0 of the etlas-cabal library
```

So I have to assume that Etlas supports cabal-version:2.0 but not yet cabal-version:2.2

Given that Eta currently isn't natively supported by plain lib:Cabal, I'm considering downgrading HsYAML-aeson to `cabal-version:2.0` if dhall-haskell still intends to support Eta

**Comment by PierreR** [comment]
> Hopefully we will build eta/etlas with a newer ghc version at some point but there is no a concrete time limit.
>if dhall-haskell still intends to support Eta
Maybe this deserves to be discussed in a separate issue given the future prospect of Eta and the amount of effort it takes for dhall to keep supporting it ?

**Comment by symbiont-rahul-muttineni** [comment]
Please don't hold back on newer Cabal features because of Etlas. It is *very* easy for us to backport Cabal features vs GHC language extensions, so go ahead and upgrade. We can workaround using our patching system temporarily and in the long-term make an effort to upgrade our fork to Cabal 2.2+

**Comment by jneira** [comment]
> Maybe this deserves to be discussed in a separate issue given the future prospect of Eta and the amount of effort it takes for dhall to keep supporting it ?

Agree, imo the pr is good  and ci is all green so it could be merged, 
thanks for your great work @vijayphoenix!

**Comment by vijayphoenix** [comment]
I suggest that we should wait till HsYAML 0.2 and HsYAML-aeson 0.2 are released

File: dhall-json/src/Dhall/Yaml.hs [hunk 6]
```diff
@@ -90,32 +94,54 @@ jsonToYaml
...
+        else defaultSchemaEncoder
 #endif
+
+
+{-# INLINE bsToStrict #-}
+bsToStrict :: Data.ByteString.Lazy.ByteString -> ByteString
+#if MIN_VERSION_bytestring(0,10,0)
+bsToStrict = Data.ByteString.Lazy.toStrict
+#else
+bsToStrict = Data.ByteString.concat Data.ByteString.Lazy.toChunks
```
Comment (sjakobi) [inline dhall-json/src/Dhall/Yaml.hs]: I don't think we need this: According to https://gitlab.haskell.org/ghc/ghc/wikis/commentary/libraries/version-history GHC-7.10.3 included bytestring-0.10.6.0, so there's no need to support anything older than that.

Comment (hvr) [reply dhall-json/src/Dhall/Yaml.hs]: To be fair, this wasn't obvious from `dhall-json.cabal` which claims to be compatible with older `bytestring` versions... :-)

**Comment by ethercrow** [comment]
Does this PR make dhall GPL?

**Comment by Gabriel439** [comment]
@ethercrow: I believe it would require the `dhall-json` and `dhall-lsp-server` packages to be relicensed to be GPL or GPL-compatible, which would be fine with me since people basically only use those packages for the executables

**Comment by patrickmn** [comment]
I'm importing dhall-json in GHCJS, and GPL is problematic for me. Any chance of creating `dhall-yaml` and importing `dhall-json`, but not the other way around?

**Comment by patrickmn** [comment]
@hvr is there any chance you'd consider relicensing under e.g. BSD3? IMO this dep would be a significant obstacle to Dhall adoption, particularly of dhall-kubernetes, since Kubernetes is typically used by companies hesitant to accept GPL deps (whether separate binaries or not.)

**Comment by Gabriel439** [comment]
Yeah, I think we will have to hold off on merging this unless `HsYAML` can be relicensed.  These aren't the only concerns I've heard about the switch to GPL

**Comment by hvr** [comment]
Fwiw, we just released

- http://hackage.haskell.org/package/HsYAML-0.2.0.0
- http://hackage.haskell.org/package/HsYAML-aeson-0.2.0.0

which incorporates the work from [Vijay Tadikamalla's GSOC 2019 Project](https://vijayphoenix.github.io/blog/gsoc-the-conclusion/).

I'm sorry to hear that the license appears to cause some worries; I'm sorry but I simply can't relicense. I'm also a bit sad to hear that a hypothetical worry that some commercial entities *might* hesitate to use GPL licensed OSS would prevent `dhall-haskell` and the GHCJS and Eta community from benefitting from all the hard work Vijay and I put into making HsYAML the best Haskell YAML library it can be. And yet some of the most successful OSS projects out there are GPL licensed; this would mean that said commercial entities would also be hesitant to use Linux systems which besides the OS kernel are made up by large amount of GPL licensed programs and libraries. Companies such as Microsoft have been vilifying the GPL over a decade ago, and yet they've come around and today they include a GPL'ed Linux systems themselves and provide their own [custom WSL Linux Kernel](https://github.com/microsoft/WSL2-Linux-Kernel). Given that Microsoft has reconsidered their stance, can't you reconsider too?

**Comment by ocharles** [comment]
Often it is not the company itself that makes that stance, it's the external stakeholders. For example, we're looking into another investment round and part of that is going to require an evaluation of software licenses we use, which may result in changes. I don't think we're unique in this situation

**Comment by patrickmn** [comment]
My own opinion: I wish corporations (~Fortune 1000) would accept GPL, especially when it's merely a question of a standalone binary. I've often made the analogy that running a GPL'ed binary is the same as running your own application on Linux -- you don't become subject to it just because of that. I wish they would treat arbitrary GPL binaries like they do Linux, or any of the "common" packages to which you refer -- but they just don't.

There is a huge difference between saying "To run this, you'll need to run apt-get install `libgmp-dev`" (something that is already "inside" and approved) and "Some of the software we're asking you to approve is GPL". It's not a *technical* difference, but it nevertheless makes a huge difference. It always, ALWAYS, kicks off several internal processes that will require exceptions if they're considered at all. Quite a different situation from e.g. Microsoft releasing a fork of Linux.

I've lost count of how many conversations I've been in that started with "So is any of the software you're shipping GPL or LGPL?", and how many times that has ended in delays/being turned down because of even LGPL'ed standalone binaries.

IMO, if you're speaking about large companies--the exact type that frequently uses Kubernetes--this is not hypothetical. It is the norm, not the exception.

**Comment by hvr** [comment]
Btw,

> "To run this, you'll need to run `apt-get install libgmp-dev`"

You can actually say "..., you'll need to run `apt-get install libghc-hsyaml-dev`" since HsYAML has been already included in [Debian](https://packages.debian.org/source/buster/haskell-hsyaml) and other Linux distros...

**Comment by f-f** [comment]
Commenting to add another concern about licensing: I love GPL and I do release some projects with it, but I think in this case the problem is about this change vs the ecosystem: if the ecosystem was already GPL then there would be no problem because everyone would be used to it. Since the ecosystem in Haskell is mostly BSD3, the hurdle to relicense everything is a thing because all the downstreams will need to relicense. As a downstream of `dhall-json` I would not want to relicense to GPL so I'd drop my usage of `dhall-json` if it'd switch to GPL

**Comment by hvr** [comment]
> ...the hurdle to relicense everything is a thing because all the downstreams will need to relicense

No it's not "a thing" because there is such need and nobody's is being asked to change their code's license! ;-)

I wonder if these misconceptions are still an echo of the anti-GPL FUD Microsoft et al. disseminated at the beginning of this century before they came around and embraced GNU/Linux, GCC, Git, MySQL, etc...

It's quite simple actually: If your BSD3 project "A" starts depending on a GPL licensed library "B", or is otherwise aggregated into a combined work with "B" (and if you think about, this may easily occur when a 3rd party decides to depend on both a BSD3 licensed library and a GPL licensed library at any given time without you even knowing about it!) your projects "A"'s code still remains BSD3 licensed. There's no need whatsoever to change your own code's BSD3 license to become GPL licensed. The actual consequence is that the *combined* work when distributed needs to meet the additional requirements of the GPL (or in general, the sum of all parts' licenses that make up the combined work); but the individual parts retain their respective individual licensing.

There's also the [GPL FAQ](https://www.gnu.org/licenses/gpl-faq.en.html) which can help debunk other GPL licensing related misconceptions.

**Comment by Gabriel439** [comment]
I don't think this GitHub pull request is an appropriate forum to have this discussion.

@hvr @vijayphoenix: I apologize that you both invested your time in this and I should have foreseen that this would have been a problem earlier.  That said, I'm still deciding not to accept a GPL dependency.

Based on that decision, there are only two possible outcomes for this pull request:

* `HsYAML` is relicensed to not be GPL

  I assume this is unlikely, but I'm mentioning it for completeness.  I don't presume that you would relicense `HsYAML` just for `dhall`

* This pull request is closed

Both outcomes are unfortunate, but I think it's better to be decisive than to let tension surrounding the choice of licenses continue to build.

**Comment by Gabriel439** [comment]
So @hvr discussed this with me out-of-band and the agreement we came to was that this would be okay if the GPL code was guarded behind a `cabal configure` flag.  The reason this does not add complexity is because this would also allow us to remove the Eta-specific code, so it evens out.

In other words, currently there are two cabal build configurations:

* Use `libyaml` - The default
* Using a Java YAML package - For Eta

... and after this change the two configurations would become:

* Use `libyaml` - Still the default
* Use `HsYAML` - For Eta or improved functionality

**Comment by patrickmn** [comment]
I'm using Eta to avoid the similar "We can't run Haskell" concern that goes hand in hand with "Is any of your code GPL?", so I think I'd have to maintain a fork for this. Tbf I'm not really sure what's going on with the Eta project (have tried reaching out several times to discuss supporting it financially, but getting no replies, and the company's website has been deregistered and squatted) so this would be the lesser concern...

**Comment by jneira** [comment]
Well, our original plan was to replace `eta` and `ghcjs` specific code, being HsYaml pure haskell, so maybe we could keep the eta one as well. I was not aware that `dhall-json` was being used compiled with eta in contexts where gpl could be problematic.

RE `eta` status: @rahulmutt talked in gitter [about it](https://gitter.im/typelead/eta?at=5d233d37198ccc7033acf9a9). Personally i want to keep contribuying everything i can.

**Comment by Gabriel439** [comment]
@jneira: So I'll propose that we go ahead with the original plan to replace the Eta bindings with HsYaml and let @patrickmn maintain an internal fork with Eta bindings like he suggested

**Comment by patrickmn** [comment]
Would you consider a PR to move the YAML code in `dhall-json` to a new package `dhall-yaml`?

I wrote a library to encode an Aeson Value as YAML: https://github.com/clovyr/aeson-yaml. Given that and a separate `dhall-yaml`, a fork of `dhall-json` wouldn't be necessary, and it wouldn't have issues with the `yaml`/`HsYaml` deps.

**Comment by Gabriel439** [comment]
@patrickmn: So I think it does make sense to factor out `dhall-to-yaml` into a separate package, but I think we should resolve this pull request first before doing that so that @vijayphoenix doesn't have to rebase their changes

**Comment by vijayphoenix** [comment]
@Gabriel439 I have updated the PR. So, you can merge it as soon as the checks at are completed

**Comment by patrickmn** [comment]
sgtm!

File: dhall-json/src/Dhall/Yaml.hs [hunk 7]
```diff
@@ -23,11 +24,12 @@ import qualified Options.Applicative
 #if defined(ETA_VERSION)
 import Dhall.Yaml.Eta ( jsonToYaml )
 #else
-import qualified Data.Yaml
-# if MIN_VERSION_yaml(0,10,2)
-import qualified Data.Text
-import qualified Text.Libyaml
-# endif
+import qualified Data.YAML.Aeson as YAML
```
Comment (sjakobi) [inline dhall-json/src/Dhall/Yaml.hs]: `YAML` seems like a rather ambiguous name here.
```suggestion
import qualified Data.YAML.Aeson
```

File: nix/ordered-containers.nix [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{ mkDerivation, base, containers, stdenv }:
+mkDerivation {
+  pname = "ordered-containers";
+  version = "0.2.2";
+  sha256 = "c75ac7330e70cd5d6ac0062b68033779cf15cd986d4ca20f838e016d466d22c9";
+  libraryHaskellDepends = [ base containers ];
+  description = "Set- and Map-like types that remember the order elements were inserted";
```
Comment (sjakobi) [inline nix/ordered-containers.nix]: That's very interesting! Do these data structures perform well? `dhall` currently uses custom types `Dhall.Map` and `Dhall.Set` for this.

**Comment by Gabriel439** [comment]
Before merging I just want to make sure I understand what we all agree on.  I think the plan is:

* I merge @vijayphoenix as it currently is once it passes all checks
* @patrickmn follows up with another change to use https://github.com/clovyr/aeson-yaml depending on the value of a `cabal configure` flag
* `yaml-to-dhall` would still be GPL because `aeson-yaml` only supports encoding, but I think people would be okay with that since `yaml-to-dhall` is typically only used interactively by humans and not as part of deployed software or build pipelines

Does that sound right to everybody?

**Comment by Gabriel439** [comment]
@vijayphoenix: To fix the Hydra build failure I think you need to run:

```bash
$ cabal2nix cabal://HsYAML-aeson > ./nix/HsYAML-aeson.nix
```

**Comment by vijayphoenix** [comment]
Done!

**Comment by Gabriel439** [comment]
@vijayphoenix: Thank you for your patience! 🙂

**Pull request merged**